### PR TITLE
audit: Port patches audit code to a rubocop and add tests

### PR DESCRIPTION
**Pull request by GauthamGoli**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----
As of now `patch_problems` method is replicated at two different places(`audit.rb` and `legacy_patches_cop.rb`).  

In #2790 I told that `patch_problems` method would be moved to `formula_cop.rb` in future, since `audit.rb` code which calls `patch_problems` had not been completely ported yet then. 
 
In this PR, I ported rest of the code, hence we can remove `patch_problems` from `audit.rb`, and move its counterpart defined in `legacy_patches_cop.rb` to `formula_cop.rb` but I'm skeptical about it as I noticed that all the methods defined in `FormulaCop` class are very generalized methods.  
Is `FormulaCop` the right place for `patch_problems`? If not, I would refactor it accordingly.  Thanks!

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
> Is FormulaCop the right place for patch_problems?

Yeh, I agree that it's probably not. Either you could have a shared class for the two classes using this or just put the external patch audits in the same class as the legacy ones.

**Comment by GauthamGoli** [comment]
I will keep external patch audits in the same class as legacy patch audits.

**Comment by JCount** [comment]
@GauthamGoli That approach seems to make the most sense. Grouping the logic of the patch audits together should hopefully help with maintenance, and perhaps coverage as well.

File: Library/Homebrew/rubocops/patches_cop.rb [hunk 1]
```diff
@@ -4,16 +4,27 @@
 module RuboCop
   module Cop
     module FormulaAudit
-      # This cop checks for and audits legacy patches in Formulae
-      class LegacyPatches < FormulaCop
+      # This cop audits patches in Formulae
+      class Patches < FormulaCop
         def audit_formula(_node, _class_node, _parent_class_node, body)
+          # Audit External Patches
+          patches = find_all_blocks(body, :patch)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/patches_cop.rb]: Could just name this `external_patches` and avoid the comments.

File: Library/Homebrew/test/rubocops/patches_cop_spec.rb [hunk 2]
```diff
@@ -84,6 +85,15 @@ def patches
                                   line: 5,
                                   column: 5,
                                   source: source }]
+        elsif patch_url =~ %r{https?://patch-diff\.githubusercontent\.com/raw/(.+)/(.+)/pull/(.+)\.(?:diff|patch)}
+          expected_offenses = [{  message:  "use GitHub pull request URLs:\n"\
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/rubocops/patches_cop_spec.rb]: Favour using `EOS.indent` over embedding `\n` in multiline strings. If you've done this in other tests: make a new PR (or add to this one) that global style change. Alternatively, these could all be addressed in another PR after this is merged?

Comment (GauthamGoli) [inline Library/Homebrew/test/rubocops/patches_cop_spec.rb]: Yeah will change these in another PR.

**Comment by MikeMcQuaid** [comment]
Thanks again @GauthamGoli!

**Comment by JCount** [comment]
Good work, once again @GauthamGoli!
